src/v/archival/service.h

```diff
@@ -0,0 +1,180 @@
+#include "storage/log_manager.h"
+#include "storage/ntp_config.h"
+#include "storage/segment.h"
+#include "storage/segment_set.h"
+#include "utils/intrusive_list_helpers.h"
+
+#include <seastar/core/abort_source.hh>
+#include <seastar/core/weak_ptr.hh>
+
+#include <absl/container/node_hash_map.h>
+
+#include <queue>
+
+namespace archival {
+namespace internal {
+
+class ntp_upload_queue {
+    struct upload_queue_item {
+        ss::lw_shared_ptr<ntp_archiver> archiver;
+        intrusive_list_hook _upl_hook;
+        intrusive_list_hook _del_hook;
+    };
+    using hash_map = absl::node_hash_map<model::ntp, upload_queue_item>;
+
+public:
+    using value = ss::lw_shared_ptr<ntp_archiver>;
+    using key = model::ntp;
+
+    ntp_upload_queue() = default;
+
+    /// Insert new item into the queue
+    bool insert(value archiver);
+    void erase(const key& ntp);
+    bool contains(const key& ntp) const;
+    value operator[](const key& ntp) const;
+    /// Copy all archiver ntp into output iterator
+    template<class FwdIt, class Func>
+    void copy_if(FwdIt it, const Func& pred) const;
+
+    /// Return number of archivers that queue contains
+    size_t size() const noexcept;
+
+    /// Get next upload candidate
+    value get_upload_candidate();
+    /// Get next delete candidate
+    value get_delete_candidate();
+
+private:
+    hash_map _archivers;
+    intrusive_list<upload_queue_item, &upload_queue_item::_upl_hook>
+      _upload_queue;
+    intrusive_list<upload_queue_item, &upload_queue_item::_del_hook>
+      _delete_queue;
+};
+
+template<class FwdIt, class Func>
+void ntp_upload_queue::copy_if(FwdIt out, const Func& pred) const {
+    for (const auto& [ntp, _] : _archivers) {
+        if (pred(ntp)) {
+            *out++ = ntp;
+        }
+    }
+}
+
+/// Shard-local archiver service.
+/// The service maintains a working set of archivers. Every archiver maintains a
+/// single partition. The working set get reconciled periodically. Unused
+/// archivers gets removed and new are added. The service runs a simple
+/// workflow on its working set:
+/// - Reconcile working set
+/// - Choose next candidate(s) archiver(s)
+/// - Start configured number of uploads
+/// - Start configured number of deletes
+/// - Re-upload manifest(s)
+/// - Reset timer
+class scheduler_service_impl {
+public:
+    /// \brief create scheduler service
+    ///
+    /// \param configuration is a archival cnfiguration
+    /// \param api is a storage api service instance
+    /// \param pm is a partition_manager service instance
+    scheduler_service_impl(
+      const configuration& conf,
+      ss::sharded<storage::api>& api,
+      ss::sharded<cluster::partition_manager>& pm);
+
+    /// \brief create scheduler service config
+    /// This mehtod will use shard-local redpanda configuration
+    /// to generate the configuration.
+    static ss::future<archival::configuration> get_archival_service_config();
+
+    /// Start archiver
+    void start();
```

Comment:
i think seastar is ok with start returning void, but it seems like a normal pattern to return ss::future<>